engine/access/rest/websockets/controller_test.go

```diff
@@ -0,0 +1,990 @@
+
+		conn.AssertExpectations(t) // Should not invoke WriteMessage after context cancellation
+	})
+}
+
+// newControllerMocks initializes mock WebSocket connection, data provider, and data provider factory.
+// The mocked functions are expected to be called in a case when a test is expected to reach WriteJSON function.
+func newControllerMocks(t *testing.T) (*connmock.WebsocketConnection, *dpmock.DataProviderFactory, *dpmock.DataProvider) {
+	conn := connmock.NewWebsocketConnection(t)
+	conn.On("Close").Return(nil).Once()
+	conn.On("SetPongHandler", mock.AnythingOfType("func(string) error")).Return(nil).Once()
+	conn.On("SetReadDeadline", mock.Anything).Return(nil)
+	conn.On("SetWriteDeadline", mock.Anything).Return(nil)
+
+	dataProvider := dpmock.NewDataProvider(t)
+	factory := dpmock.NewDataProviderFactory(t)
+
+	return conn, factory, dataProvider
+}
+
+// expectSubscribeRequest mocks the client's subscription request.
+func (s *WsControllerSuite) expectSubscribeRequest(t *testing.T, conn *connmock.WebsocketConnection) string {
+	request := models.SubscribeMessageRequest{
+		BaseMessageRequest: models.BaseMessageRequest{
+			MessageID: uuid.New().String(),
+			Action:    models.SubscribeAction,
+		},
+		Topic: "blocks",
+	}
+	requestJson, err := json.Marshal(request)
+	require.NoError(t, err)
+
+	// The very first message from a client is a request to subscribe to some topic
+	conn.
+		On("ReadJSON", mock.Anything).
+		Run(func(args mock.Arguments) {
+			msg, ok := args.Get(0).(*json.RawMessage)
+			require.True(t, ok)
+			*msg = requestJson
+		}).
+		Return(nil).
+		Once()
+
+	return request.MessageID
+}
+
+// expectSubscribeResponse mocks the subscription response sent to the client.
+func (s *WsControllerSuite) expectSubscribeResponse(t *testing.T, conn *connmock.WebsocketConnection, msgId string) {
+	conn.
+		On("WriteJSON", mock.Anything).
+		Run(func(args mock.Arguments) {
+			response, ok := args.Get(0).(models.SubscribeMessageResponse)
+			require.True(t, ok)
+			require.Equal(t, msgId, response.MessageID)
+			require.Equal(t, true, response.Success)
+		}).
+		Return(nil).
+		Once()
+}
+
+func (s *WsControllerSuite) expectCloseConnection(conn *connmock.WebsocketConnection, done <-chan struct{}) {
+	// In the default case, no further communication is expected from the client.
+	// We wait for the writer routine to signal completion, allowing us to close the connection gracefully
+	conn.
+		On("ReadJSON", mock.Anything).
+		Return(func(msg interface{}) error {
+			for range done {
+			}
+			return websocket.ErrCloseSent
+		}).
+		Once()
+
+	s.expectKeepaliveRoutineShutdown(conn, done)
+}
+
+func (s *WsControllerSuite) expectKeepaliveRoutineShutdown(conn *connmock.WebsocketConnection, done <-chan struct{}) {
+	// We use Maybe() because a test may finish faster than keepalive routine trigger WriteControl
+	conn.
+		On("WriteControl", websocket.PingMessage, mock.Anything).
+		Return(func(int, time.Time) error {
+			for {
```

Comment:
doesn't look like this needs a loop since it's guaranteed to return. is that intended?